# Prediction of m6A Reader substrate sites using deep convolutional and recurrent neural network  
## Abstract
N6-methyladenosine (m6A) is the most prevalent post-transcriptional modification in mRNA since it could regulate some significant biological functions with the binding of some m6A reader proteins. Multiple readers exist in the human genome, however, the binding specificity was not clarified due to the limited wet experiments on this topic. Therefore, we devised a deep learning approach which incorporated CNN and RNN frameworks together to predict the epitranscriptome-wide targets of six m6A reader proteins (YTHDF1-3, YTHDC1-2, EIF3A). We also utilized layer-wise relevance calculation to obtain contribution of each input feature. Our model achieved state-of-the-art performance with the average AUROC of 0.942 in EIF3A full transcript, compared with 0.929 in CNN-only framework and 0.817 in Support Vector Machine (SVM) method under same condition. Besides, we identified the optimal sequence length (1001bp) in the m6A reader substrate prediction. The results provide new insight into epitranscriptome target prediction and functional characterization of m6A readers.

## Criteria of determining the m6A reader binding sites 
![Alt text](https://github.com/yuxuanwu17/m6A_reader/blob/master/plot/m6a_criteria.png) 

**Figure 1** Criteria of determining the m6A reader binding sites. There are three significant factors to consider when determining the m6A reader binding sites, which is DRACH motif, known m6A Sites and CLIP labeled Sites. Generally, m6A readers have a tendency to bind known m6A sites and DRACH motif, but these two are not robust enough, without CLIP labeled sites, they could still be considered as negative samples. Therefore, in this experiment, three factors have to be satisfied simultaneously. In addition, the position of gene model would not influence the samples.

## Overall architecture of the deep learning model
![Alt text](https://github.com/yuxuanwu17/m6A_reader/blob/master/plot/Architecture(3).png) 

**Figure 2** The sequence data are encoded by One-hot method and fed into the convolution layer and followed by the pooling layer twice to extract the significant features. The LSTM layer learns the long-term dependencies between sequence data generated by convolution layers. The flatten layer combines the previous kernels into a vector and inputs to the fully connected layer to calculate the probability of being m6A reader substrate site 

## Prerequisite packages
keras version 2.3.0, numpy, pandas, argparse

## Run 'python3 main.py -h' in command line for help

### Select the input genes
You could select 6 genes, including YTHDC1-2, YTHDF1-3, EIF3A

### Select the input length of the sequence
You could choose 251, 501, 1001, 2001bp input length to compare the prediction performance. 
Due to the size limitation, i only upload some some small sized sequence as examples.
Email me, if you want to apply all the data

### Select the condition of the seqeunce, either full transcript (full) or mature RNA (exon)
To minimize the bias in selecting the polyA RNAs, we prepared the full transcript data and mature RNA data. In detail, mature RNA data exclude the sites on the intron region while the full transcript data covered either the exon or intron region.

### Choose the CNN or CNN+RNN model
To compare the performance under different framework

## Results

### Performance comparison 
![Alt text](https://github.com/yuxuanwu17/m6A_reader/blob/master/plot/cmbd2.png)

**Figure 3** (A) Compared the performance of CNN model and CNN+RNN model in the prediction of six m6A reader substrates under different length in full transcripts. (B) Compared the AUROC value in either full transcript or mature transcript when predicting the EIF3A reader substrates. 

### ROC and PR-curve comparison between multiple sequence length

![Alt text](https://github.com/yuxuanwu17/m6A_reader/blob/master/plot/rocs.png)

**Figure 4** (A) Compared the ROC curve and the regarding AUROC of mature and full transcript of EIF3A under CNN+RNN model in various sequence lengths. (B) Compared the PR curve and the regarding PRAUC of mature and full transcript of EIF3A under CNN+RNN model in various sequence lengths. 

### Quantify each input nucleotide contribution by the layer-wise relevance calculation
![Alt text](https://github.com/yuxuanwu17/m6A_reader/blob/master/plot/contribution_plot.png)

**Figure 5** Feature importance scores in EIF3A full transcript prediction. We both extracted upstream/downstream 50bp and upstream/downstream 250 bp of the sequence to rank the contribution of each nucleotide in determining the binding site. In each position, the higher score it gains, the higher contribution towards the binding sites.
